pkg/proctree/proctree.go

```diff
@@ -0,0 +1,113 @@
+package proctree
+
+import (
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/aquasecurity/tracee/pkg/errfmt"
+)
+
+//
+// The way the process tree is organized:
+//
+// 1. All processes and thread group leaders are stored as "processes" (hash -> process).
+// 2. All threads are stored as "threads" (hash -> thread).
+// 3. Each process has a list of children (hashes of real forked processes).
+// 4. Each process, that is a thread group leader, has a list of threads (hashes of threads).
+// 5. Threads from the same thread group have the same parent, the parent of the group leader.
+// 6. Each thread has a parent (hash of parent) and a leader (hash of thread group leader).
+//
+// Instruction:
+//
+// 1. To add an artifact to a process, simply add it to the process entry using its hash.
+// 2. To add an artifact to a process parent, use the process parent hash.
+// 3. To add an artifact to a thread, pick the group leader hash, add it to its process entry.
+//
+// NOTE: The importance of having the thread group leader hash to each thread is the following: the
+// thread group leader is considered a "Process", in the proctree, and it will have all the
+// information regarding the process, such as the executable and interpreter. Whenever an artifact
+// log, for a thread, is needed, the thread group leader hash will be used to find the process so it
+// can be logged as a process artifact.
+//
+
+const proctreeCacheSize = 65536 // 64K (should be enough for most use cases)
```

Comment:
I think this entry should be configurable since different environments may require different cache sizes.
Also, since an lru is used - how do we fill back an entry that was evicted from the cache (and we now see an event of this process/thread)?

Reply:
My idea for filling back was keeping the events from the regular pipeline accessing the procinfo (so the cache would be hot for them). We could fill the process tree from a regular event if the process/thread does not exist, for example.

About the cache size, sounds good to me having it configurable.

Reply:
Actually it might be a good idea to complete the missing data from procfs lazily.
If we are missing a process or thread - update it through procfs.
By doing this, we can also remove the need to initialize from procfs at init time (we can also keep it, but we need to keep in mind that there might be race conditions between the time we update from procfs and until we start getting the signals)